core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -593,9 +722,147 @@ private Builder visitAggregate(Aggregate e, List<Integer> groupKeyList,
+
+    // Extract the alias of the subQuery
+    String subQueryAlias = extractSubQueryAlias(sqlBasicCall);
+
+    // Identify column names used from the subQuery in the GROUP BY clause
+    List<String> columnNamesUsedFromSubQuery = extractColumnNamesUsedFromSubQuery(groupByList, subQueryAlias);
+
+    // Create a list of SqlIdentifier for the identified column names
+    List<SqlIdentifier> sqlIdentifiersNew = createSqlIdentifiers(columnNamesUsedFromSubQuery);
+
+    // Update the SELECT list in the subQuery
+    updateSelectListInSubQuery(sqlBasicCall, sqlIdentifiersNew);
+  }
+
+  // Function to check if the FROM clause is a join with a basic call
+  private boolean isJoinWithBasicCall(Builder builder) {
+    return builder.select.getFrom() instanceof SqlJoin &&
+        ((SqlJoin) builder.select.getFrom()).getLeft() instanceof SqlBasicCall;
+  }
+
+  // Function to get the left operand if it is a basic call
+  private SqlBasicCall getLeftBasicCall(Builder builder) {
+    return (SqlBasicCall) ((SqlJoin) builder.select.getFrom()).getLeft();
+  }
+
+  // Function to extract the alias of the subquery
+  private String extractSubQueryAlias(SqlBasicCall sqlBasicCall) {
+    String subQueryAlias = null;
+    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
+      if (sqlNode instanceof SqlIdentifier) {
+        subQueryAlias = ((SqlIdentifier) sqlNode).names.get(0);
+      }
+    }
+    return subQueryAlias;
+  }
+
+  // Function to extract column names used from the subQuery in the GROUP BY clause
+  private List<String> extractColumnNamesUsedFromSubQuery(List<SqlNode> groupByList, String subQueryAlias) {
+    List<String> columnNamesUsedFromSubQuery = new ArrayList<>();
+    for (SqlNode groupByNode : groupByList) {
+      if (groupByNode instanceof SqlIdentifier) {
+        if (((SqlIdentifier) groupByNode).names.get(0).equals(subQueryAlias)) {
+          columnNamesUsedFromSubQuery.add(((SqlIdentifier) groupByNode).names.get(1));
+        }
+      }
+    }
+    return columnNamesUsedFromSubQuery;
+  }
+
+  // Function to create a list of SqlIdentifier for the identified column names
+  private List<SqlIdentifier> createSqlIdentifiers(List<String> columnNamesUsedFromSubQuery) {
+    List<SqlIdentifier> sqlIdentifierList = new ArrayList<>();
+    for (String col : columnNamesUsedFromSubQuery) {
+      sqlIdentifierList.add(new SqlIdentifier(ImmutableList.of(col), POS));
+    }
+    return sqlIdentifierList;
+  }
+
+  // Function to update the SELECT list in the subQuery
+  private void updateSelectListInSubQuery(
+      SqlBasicCall sqlBasicCall, List<SqlIdentifier> sqlIdentifiersNew) {
+    List<String> columnNamesInSelect = new ArrayList<>();
+    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
+      if (sqlNode instanceof SqlSelect) {
+        SqlNodeList sqlNodeList = ((SqlSelect) sqlNode).getSelectList();
+        if (sqlNodeList != null) {
+          for (SqlNode selectNode : sqlNodeList) {
```

Comment:
this code looks complex can we do refactoring by the way...

private void updateSelectListInSubQuery(SqlBasicCall sqlBasicCall, List<SqlIdentifier> sqlIdentifiersNew) {
    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
        if (sqlNode instanceof SqlSelect) {
            updateSelectListInSelectNode((SqlSelect) sqlNode, sqlIdentifiersNew);
        }
    }
}

private void updateSelectListInSelectNode(SqlSelect sqlSelect, List<SqlIdentifier> sqlIdentifiersNew) {
    SqlNodeList sqlNodeList = sqlSelect.getSelectList();

    if (sqlNodeList != null) {
        List<String> columnNamesInSelect = getColumnNamesInSelect(sqlNodeList);

        for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
            if (!columnNamesInSelect.contains(sqlIdentifier.names.get(0))) {
                sqlNodeList.add(createNewSelectItem(sqlIdentifier, sqlSelect));
            }
        }

        sqlSelect.setSelectList(new SqlNodeList(sqlNodeList, POS));
    } else if (sqlIdentifiersNew.size() > 0) {
        updateSelectListForJoin(sqlSelect, sqlIdentifiersNew);
    }
}

private List<String> getColumnNamesInSelect(SqlNodeList sqlNodeList) {
    List<String> columnNamesInSelect = new ArrayList<>();

    for (SqlNode selectNode : sqlNodeList) {
        if (selectNode instanceof SqlIdentifier) {
            SqlIdentifier identifier = (SqlIdentifier) selectNode;
            columnNamesInSelect.add(identifier.names.size() > 1 ? identifier.names.get(1) : identifier.names.get(0));
        }
    }

    return columnNamesInSelect;
}

private void updateSelectListForJoin(SqlSelect sqlSelect, List<SqlIdentifier> sqlIdentifiersNew) {
    if (sqlSelect.getFrom() instanceof SqlJoin) {
        SqlNodeList sqlNodes = new SqlNodeList(POS);

        for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
            sqlNodes.add(createNewSelectItem(sqlIdentifier, sqlSelect));
        }

        sqlSelect.setSelectList(new SqlNodeList(sqlNodes, POS));
    }
}

Reply:
Done